File: DKF-Python/ComputeWeightMats.py
import numpy as np
import scipy.io as sio

def ComputeWeightMats(Directory,NetParameters):
    """
    Computes the weight matrices used for residue scaling.

    Parameters:
        NetParameters (dict): Dictionary containing network parameters.

    Returns:
        MeasurementWeightMats (list of numpy.ndarray): Measurement weight matrices for each layer.
        PredictorWeightMats (list of numpy.ndarray): Predictor weight matrices for each layer.
        MeasurementWeightMatsSym (list of numpy.ndarray, optional): Symmetric measurement weight matrices.
        PredictorWeightMatsSym (list of numpy.ndarray, optional): Symmetric predictor weight matrices.
    """
    Layers = NetParameters['Layers']
    StateDimension = NetParameters['StateDimension']
    ObservationDimension = NetParameters['ObservationDimension']
    WeightMats = NetParameters['WeightMats']

    MeasurementWeightMats = [None] * Layers
    PredictorWeightMats = [None] * Layers

    # Cleaner propagation for test phase
    MeasurementWeightMatsSym = [None] * Layers
    PredictorWeightMatsSym = [None] * Layers

    # Compute weight matrices
    if WeightMats == 'Identity':
        for Layer in range(Layers):
            MeasurementWeightMats[Layer] = np.eye(ObservationDimension)
            PredictorWeightMats[Layer] = np.eye(StateDimension)

            MeasurementWeightMatsSym[Layer] = np.eye(ObservationDimension)
            PredictorWeightMatsSym[Layer] = np.eye(StateDimension)

    elif WeightMats == 'Input':
        Experiment = NetParameters['Experiment']

        MeasurementWeightMats = sio.loadmat(Directory+f'MeasurementWeightMatsExp{Experiment}.mat',squeeze_me = True)['MeasurementWeightMats']
        PredictorWeightMats = sio.loadmat(Directory+f'PredictorWeightMatsExp{Experiment}.mat',squeeze_me = True)['PredictorWeightMats']

        for Layer in range(Layers):
            MeasurementWeightMatsSym[Layer] = 0.5 * (MeasurementWeightMats[Layer] + MeasurementWeightMats[Layer].T)
            PredictorWeightMatsSym[Layer] = 0.5 * (PredictorWeightMats[Layer] + PredictorWeightMats[Layer].T)

    return MeasurementWeightMats, PredictorWeightMats, MeasurementWeightMatsSym, PredictorWeightMatsSym

File: DKF-MATLAB/ComputeWeightMats.m
function [MeasurementWeightMats,PredictorWeightMats,MeasurementWeightMatsSym,PredictorWeightMatsSym] = ComputeWeightMats(NetParameters)
%COMPUTEWEIGHTMATS: Compute the weight matrices used for residue scaling

%Variables
Layers = NetParameters.Layers;
StateDimension = NetParameters.StateDimension;
ObservationDimension = NetParameters.ObservationDimension;
WeightMats = NetParameters.WeightMats;

MeasurementWeightMats = cell(1,Layers);
PredictorWeightMats = cell(1,Layers);

%Cleaner propagation for test phase
if nargout > 2
    MeasurementWeightMatsSym = cell(1,Layers);
    PredictorWeightMatsSym = cell(1,Layers);
end

%Compute weight matrices
if strcmp(WeightMats,'Identity')
    for Layer = 1:Layers
        MeasurementWeightMats{Layer} = eye(ObservationDimension);
        PredictorWeightMats{Layer} = eye(StateDimension);

        %Cleaner propagation for test phase
        if nargout > 2
            MeasurementWeightMatsSym{Layer} = eye(ObservationDimension);
            PredictorWeightMatsSym{Layer} = eye(StateDimension);
        end
    end
end

if strcmp(WeightMats,'Input')
    Experiment = NetParameters.Experiment;

    try
        MeasurementWeightMats = importdata(['MeasurementWeightMatsExp',Experiment,'.mat']);
    catch ME
        if (strcmp(ME.identifier,'MATLAB:importdata:FileNotFound'))
            msg = ['File "MeasurementWeightMats',Experiment,'.mat" not found in current path.'];
            causeException = MException('MATLAB:myCode:FileNotFound',msg);
             ME = addCause(ME,causeException);
        end
        rethrow(ME)
    end 

    try
        PredictorWeightMats = importdata(['PredictorWeightMatsExp',Experiment,'.mat']);
    catch ME
        if (strcmp(ME.identifier,'MATLAB:importdata:FileNotFound'))
            msg = ['File "MeasurementWeightMats',Experiment,'.mat" not found in current path.'];
            causeException = MException('MATLAB:myCode:FileNotFound',msg);
             ME = addCause(ME,causeException);
        end
        rethrow(ME)
    end 

    %Cleaner propagation for test phase
    if nargout > 2
        for Layer = 1:Layers
            MeasurementWeightMatsSym{Layer} = (1/2)*( MeasurementWeightMats{Layer} + MeasurementWeightMats{Layer}' );
            PredictorWeightMatsSym{Layer} = (1/2)*( PredictorWeightMats{Layer} + PredictorWeightMats{Layer}' );
        end
    end
end


end
